korge/src/commonMain/kotlin/com/soywiz/korge3d/animation/PlaybackPattern.kt

```diff
@@ -0,0 +1,41 @@
+package com.soywiz.korge3d.animation
+
+/**
+ * Describes how the animation progress changes over time. Defines what animation frames/states are requested as the
+ * time flows. See the docs of [NormalizedTime] and [NormalizedProgress] for details.
+ */
+inline class PlaybackPattern(val value: (NormalizedTime) -> NormalizedProgress) {
```

Comment:
> This looks a lot like the Easing functionality from korma: https://github.com/korlibs/korma/blob/fb7c2de33571795c69d606b59c1ed0e1f12ad83d/korma/src/commonMain/kotlin/com/soywiz/korma/interpolation/Easing.kt#L11-L12
> 
> could we somehow reuse that?

I didn't know about easings from korma, thanks!
After a quick analysis, I'm afraid  we shouldn't use `Easing` in place of `PlaybackPattern` literally. I see two reasons:

1. **I think** easing's values are defined for arguments between 0 and 1 - could you confirm it? I'm inferring it from some easings' definitions where e.g. values above 1 would make little sense. For playback patterns, on the other hand, my intent is to define it for values from 0 to infinity, because the time flow of animations can exceed one period (as written in the doc comment).
2. Easing's API in terms of types is pretty generic (from `Double` to `Double`), while for `PlaybackPattern` I want to type it so that it's self-documenting and closer to the use case of using it for animations (from `NormalizedTime` to `NormalizedProgress`).

That's why I recommend adding an extension function to create `PlaybackPattern` out of `Easing`. It will allow reusing the multitude of existing easings and preserve nice types on `Animation3D`'s side. One problem to solve: what value should a playback pattern return for time > 1, where easing is not defined. My first thought is to provide an argument to the extension function: "residual value", so the user can define what value the animation progress is at after the easing completes. We can even take the topic of bridging `PlaybackPattern` and `Easing` into a separate PR.

What do you think?

Reply:
> If you have to change korma and korge at the same time, maybe you can make this PR into korge-next instead. In fact I would prefer PRs there so I don't have to manually copy stuff there later and resolve conflicts myself

Sure - as soon as we agree on the next steps after you read my comments, I'll post a new revision as a PR to korge-next :+1:

Reply:
1. I think that despite normally you would use 0 and 1 for the input, you can use values outside that range. Regarding to outputs, some elastic values are providing values outside 0 and 1. So I believe it is flexible regarding to that
2. Easing was not typed, because it was there probably since a lot time ago when inline classes were not even in the roadmap and was not reasonable to have such GC pressure because of that.
But we could try to type it incrementally, maybe using default methods for the interface?

If the naming doesn't match, or do you believe it is not consistent, that could be solved with a type alias?
What I want to achieve is to enrich/evolve a single API if possible in the most general place that is reasonable (like korma) so most people can benefit from this nice functionality.

Reply:
Ok, got it - typing Easing sounds like a reasonable path forward. Would you be OK with implementing playback patterns with `(Double) -> Double` for now (for simplicity, to expose this feature somehow and to be able to use it with the current implementation of Easing), and tackle typing of Easing for the sake of readability and type-safety in follow-up PRs?

Reply:
We are trying to preserve source-code compatibility (not binary) between minor versions, except for the experimental packages and APIs as much as possible.

Sure, that sounds reasonable to me. Also now that you have all the relevant information available, I'll let you decide how to proceed with the current and next PRs. 👍